### PR TITLE
Superpmi collection for real-world benchmarks

### PR DESCRIPTION
**Pull request by kunalspathak**
Add superpmi collection for real-world benchmarks present in https://github.com/dotnet/performance/tree/main/src/benchmarks/real-world:
- FSharp
- ML.NET
- ImageSharp
- Roslyn
- Physics engine
- ILLink

Successful collection: https://dev.azure.com/dnceng/internal/_build/results?buildId=2193577&view=results

### CODE REVIEW

**Comment by kunalspathak** [comment]
@dotnet/jit-contrib

File: src/coreclr/scripts/superpmi_benchmarks.py [hunk 1]
```diff
@@ -192,11 +206,25 @@ def build_and_run(coreclr_args, output_mch_name):
...
+                            " DOTNET_ZapDisable:1  DOTNET_ReadyToRun:0 " \
+                            "--iterationCount 1 --warmupCount 0 --invocationCount 1 --unrollFactor 1 --strategy ColdStart --logBuildOutput"
+    elif benchmark_binary.lower().startswith("demobenchmarks"):
+        # Disable ReadyToRun so we always JIT R2R methods and collect them
+        collection_command = f"{dotnet_exe} {benchmarks_dll}  -f *CollisionBatcherTaskBenchmarks.* *GroupedCollisionTesterBenchmarks.* *GatherScatterBenchmarks.* " \
+                            " *OneBodyConstraintBenchmarks.* *TwoBodyConstraintBenchmarks.* *ThreeBodyConstraintBenchmarks.* *FourBodyConstraintBenchmarks.* " \
+                            " *SweepBenchmarks.* *ShapeRayBenchmarks.* *ShapePileBenchmark.* *RagdollTubeBenchmark.* " \
+                            f" --corerun {os.path.join(core_root, corerun_exe)} --envVars DOTNET_JitName:{shim_name} " \
+                            " DOTNET_ZapDisable:1  DOTNET_ReadyToRun:0 " \
+                            "--iterationCount 1 --warmupCount 0 --invocationCount 1 --unrollFactor 1 --strategy ColdStart --logBuildOutput"
```
Comment (AndyAyersMS) [inline src/coreclr/scripts/superpmi_benchmarks.py:222]: How long would it take to run these normally (no BDN strategy alterations)?

**Review by AndyAyersMS — APPROVED** [review]
Seems like parts of this scripting are getting a little unwieldy, but I suppose we're not likely to add another category of benchmarks this way, so maybe there's no benefit in abstracting this further.

I would like to see us run the benchmarks more naturally (and eventually with mostly release bits) but that can wait for a subsequent change.

File: src/coreclr/scripts/superpmi_benchmarks.py (hunk 1 above)
Comment (kunalspathak) [reply src/coreclr/scripts/superpmi_benchmarks.py:222]: Thanks @LoopedBard3 for getting these numbers:

```
    ML.NET ~ 12 min
    FSharp ~8 min
    Bepuphysics ~6.5 min
    ImageSharp ~6 min
    Illink ~16.5 min
    Roslyn ~ 17 mins.
```

Comment (AndyAyersMS) [reply src/coreclr/scripts/superpmi_benchmarks.py:222]: So 56 mins.

In the job you link to above the entire thing took 49 minutes on windows, so the cost above doesn't seem too bad?

image

Comment (kunalspathak) [reply src/coreclr/scripts/superpmi_benchmarks.py:222]: The 49 mins includes the queueing time. The actual execution with `invocationCount 1` is couple of minutes.

**Comment by kunalspathak** [comment]
> parts of this scripting are getting a little unwieldy

I agree. I was thinking of just make them run using normal `benchmark` collection, but that would have still needed special handling for these benchmarks because they are in different folder and different benchmark binary names. Also, I didn't want to  include them in our regular benchmarks collection.

File: src/coreclr/scripts/superpmi_benchmarks.py (hunk 1 above)
Comment (kunalspathak) [reply src/coreclr/scripts/superpmi_benchmarks.py:222]: `-IterationCount 1`:

image


default args, so definitely the collection part slows it down further. I will merge this PR as-is.

image

**Pull request merged**